[1단계 - 출석 구현] 하루(구은선) 미션 제출합니다.

eunseongu opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [ ] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3943


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

 기능을 구현하는 것에 집중하다보니 객체지향 생활체조 요구사항의 우선순위가 낮아졌던 것 같습니다.  "getter 사용하지 않기", "indent 1로 유지하기"등의 요구사항은 구현하면서 바로 지키기가 어려웠습니다. 그래서 구현을 완료한 후에 리팩토링에 반영하려 하였으나 시간 부족으로 리팩토링하지 못하여 아쉽습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요! 7기 하루입니다. 😊

이번 미션에서는 데이터를 저장하는 구조 설계에서 많은 고민을 했습니다.
크루들의 모든 정보를 관리하며, 이름으로 조회하기 쉽도록 하기 위해 Map을 사용했습니다.
`AttendanceManager` 객체에서는 모든 정보를 가지는 `Map<String, Records> crews`를 멤버로 가지도록 구현했습니다. 각 크루들이 가진 출석 정보나 출석 상태에 대한 조회도 필요했기 때문에, `Records` 객체는 날짜를 key로 가지는 `Map<LocalDate, TimeAndStatus> records`를 멤버로 가지도록 했습니다. `TimeAndStatus`는 시간과 상태(제적, 경고, 면담)을 멤버로 가진 객체로 구현했습니다.

이렇게 구조를 설계하니 조회 하는 것은 간편해졌으나 한 크루의 상태만 출력하고 싶은 경우에도 항상 여러 객체를 거쳐 값을 반환받고, 최종적으로 Controller가 View에게 넘겨주고 있는데, 이 방식이 괜찮은 것인가에 대한 의문이 들었습니다.
보완하면 좋을 방식이 있다면 리뷰에 남겨주세요! 

혹시 이와 관련하여 공부해보면 좋을 키워드를 알려주신다면 공부해보도록 하겠습니다!
감사합니다.

JO-eusan committed:
docs: 기능 구현 목록 작성

JO-eusan committed:
docs: 목표 수정

JO-eusan committed:
docs: 출석 기록 파일 추가

JO-eusan committed:
docs: 리드미 수정

JO-eusan committed:
docs: 출석 기록 검색 예외 추가

JO-eusan committed:
test: 출석부에 정보 저장 및 수정 안내

JO-eusan committed:
refactor: 크루 데이터 사용을 위한 리포지토리 생성

JO-eusan committed:
refactor: 크루 데이터 사용을 위한 리포지토리 생성

JO-eusan committed:
feat: 출석 기능 구현

JO-eusan committed:
test: 출석 정보 변경

JO-eusan committed:
feat: 출석 정보 수정 로직 구현

JO-eusan committed:
refactor: 출석 정보 저장 구조 변경

JO-eusan committed:
refactor: 데이터 저장 방식 변경 및 불필요한 객체 제거

JO-eusan committed:
test: 출석 상태 변경

JO-eusan committed:
refactor: 출석 상태 변경 테스트 통과를 위한 수정

JO-eusan committed:
refactor: 테스트 코드 필드 생성 및 컨벤션 적용

JO-eusan committed:
refactor: 출석 수정 및 출력 방식 변경

JO-eusan committed:
test: 출석 기록 조회 및 통계 계산

JO-eusan committed:
feat: 출석 통계 로직 구현

JO-eusan committed:
test: 경고 및 면담 기준 계산

JO-eusan committed:
feat: 경고 및 제적 확인 로직 구현

JO-eusan committed:
feat: 출석 기록 확인 결과 출력

JO-eusan committed:
test: 제적 위험자 파악

JO-eusan committed:
feat: 출석 기록을 바탕으로 제적 위험자 계산

JO-eusan committed:
refactor: 패널티 계산 중복 제거

JO-eusan committed:
test: 제적 위험자 정렬

JO-eusan committed:
feat: 제적 위험자 정렬 구현

JO-eusan committed:
feat: 제적 위험자 출력

JO-eusan committed:
refactor: 오늘 날짜 계산 로직 분리

JO-eusan committed:
feat: 휴일 및 주말 처리

JO-eusan committed:
feat: 기능 선택 로직 구현

JO-eusan committed:
refactor: 예외 처리 방식 변경 및 불필요한 메서드 제거

JO-eusan committed:
refactor: 자바 컨벤션 적용

asebn1 commented on README.md:
```diff
@@ -1,3 +1,71 @@
+- [x] 변경 등교 시간을 입력받는다.
+  - [ ] 시간 형식이 유효하지 않은 경우 예외 발생
+  - [ ] 캠퍼스 운영 시간이 아닌 경우 예외 발생
+- [x] 수정 후에는 변경 전과 변경 후의 출석 기록을 출력한다. (`변경전` -> `변경후`)
```
체크리스트로 잘 작성해주셨네요~ 💯
체크되지 않은 부분도 확인해볼까요?

eunseongu replied:
모두 반영하여 수정했습니다!

asebn1 commented on src/main/java/controller/FileController.java:
```diff
@@ -0,0 +1,59 @@
+public class FileController {
+
+    private final String FILE_ERROR_MESSAGE = "파일을 불러오는 데 문제가 발생했습니다. 잠시 후 다시 시도해 주세요.";
+    private AttendanceManager attendanceManager;
```
상수는 어떻게 정의하면 좋을까요?

asebn1 commented on src/main/java/controller/FileController.java:
```diff
@@ -0,0 +1,59 @@
+            String[] attr = attendaceScanner.nextLine().split(",");
+            String name = attr[0];
+
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
매번 같은 `formatter`가 생성될 것 같아요. `while` 밖으로 이동해봅시다

asebn1 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,145 @@
+                System.out.println("유효하지 않은 번호입니다.");
+            }
+        } while (!functionNumber.equals("q"));
+    }
```
새로운 타입이 들어오면 새로운 if문이 생길 수 있을 것 같아요.
개선할 수 있을까요?

- switch-case도 고민해볼까요?
- 메서드에 인덴트 1, 메서드 10줄을 초과하고 있어요

eunseongu replied:
switch-case로 개선하니 indent를 줄일 수 있었고, 가독성이 더 좋아진 것을 느꼈습니다.

다만 리팩터링 이후 프로그래밍 요구사항에 “else 예약어를 쓰지 말라고 하니 switch/case로 구현하는 경우가 있는데 switch/case도 허용하지 않는다.”라는 내용이 있는 것을 발견했습니다.

그래서 제로가 생각하는 switch-case의 장점은 무엇인지 궁금합니다!

asebn1 replied:
- `switch-case`도 사용할 수 있는 케이스여서 고민이 가능할 것 같다는 의미였습니다~ `switch/case도 허용하지 않는다` 부분은 제가 놓쳤던 부분이네요 ㅎㅎ
- 커맨드 패턴도 한번 학습해볼까요? 커맨드 패턴은 무엇일까요? (반영하라는 의미는 아닙니다. 공부하면 좋을 부분인 것 같아서 언급드려요)

asebn1 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,145 @@
+
+public class AttendanceController {
+
+    private final int DEFAULT_YEAR = 2024;
+    private final int DEFAULT_MONTH = 12;
+    private final int DEFAULT_DAY = 13;
+    private final String INPUT_DATE_FORMAT = "%04d-%02d-%02d";
+    private final String INPUT_TIME_FORMAT = "%02d:%02d";
+    private final InputView inputView;
+    private final OutputView outputView;
+    private final AttendanceManager attendanceManager;
```
상수는 static을 붙여줍시다 :)
```suggestion
    private final int DEFAULT_YEAR = 2024;
    private final int DEFAULT_MONTH = 12;
    private final int DEFAULT_DAY = 13;
    private final String INPUT_DATE_FORMAT = "%04d-%02d-%02d";
    private final String INPUT_TIME_FORMAT = "%02d:%02d";
    
    private final InputView inputView;
    private final OutputView outputView;
    private final AttendanceManager attendanceManager;
```

그리고 상수와 일반 변수 사이엔 개행을 넣는 편이에요

asebn1 commented on src/main/java/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,67 @@
+    public Map<String, StatisticsResult> sortCrew(LocalDate nowDate) {
+        Map<String, StatisticsResult> sortedResult = findWarningCrews(nowDate);
+
+        return sortedResult
+            .entrySet().stream().sorted(Comparator.comparing(
+                    (Map.Entry<String, StatisticsResult> entry) -> entry.getValue().getPenalty(),
+                    Comparator.naturalOrder())
+                .thenComparing(entry -> entry.getValue().getAbsenceCount(),
+                    Comparator.reverseOrder()).thenComparing(Map.Entry::getKey)
+            )
+            .collect(LinkedHashMap::new, (map, entry) -> map.put(entry.getKey(), entry.getValue()),
+                Map::putAll);
```
참고하면 좋을 자료 공유드립니다

- [스트림은 주의해서 사용하라](https://velog.io/@semi-cloud/Effective-Java-%EC%95%84%EC%9D%B4%ED%85%9C-45-%EC%8A%A4%ED%8A%B8%EB%A6%BC%EC%9D%80-%EC%A3%BC%EC%9D%98%ED%95%B4%EC%84%9C-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC)

asebn1 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,145 @@
+        } while (!functionNumber.equals("q"));
+    }
+
+    private void attend() {
+        try {
+            String name = inputView.readName();
+            attendanceManager.findByName(name);
+            List<String> time = List.of(inputView.readTime().split(":"));
+
+            String dateForm = String.format(INPUT_DATE_FORMAT, DEFAULT_YEAR, DEFAULT_MONTH,
+                DEFAULT_DAY);
+            String timeForm = formatTime(time);
+            LocalDateTime dateTime = formatDateTime(dateForm, timeForm);
+
+            TimeAndStatus timeAndStatus = attendanceManager.attendCrew(name, dateTime);
+            outputView.printAttendanceRecord(dateTime.toLocalDate(), timeAndStatus);
+        } catch (IllegalArgumentException e) {
+            System.out.println(e.getMessage());
+        }
+    }
```
- `":"`도 상수화가 가능할 것 같아요
- `attendanceManager.findByName(name)`의 목적은 무엇인가요?

eunseongu replied:
controller에서 크루의 닉네임이 존재하는지 여부에 대한 검증을 위한 로직이 필요했는데, 이미 같은 역할을 하는 findByName 메서드가 존재하여 사용했습니다.
다만 지금보니 controller에서는 반환값이 필요없기 때문에 검증하는 메서드를 따로 생성하던지, 혹은 반환값을 사용하도록 로직을 변경하는 것이 좋을 것 같습니다.

리팩터링을 통해 반환값이 없는 `attendanceManager.hasCrew`를 사용하여 입력된 닉네임이 존재하는지 검증하도록 처리했습니다.

asebn1 commented on src/main/java/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,67 @@
+        return AttendanceStatistics.calculateExpelledWarning(nowDate, crews);
+    }
+
+    public Records findByName(String name) {
+        if (crews.get(name) == null) {
+            throw new IllegalArgumentException("존재하지 않는 크루입니다.");
+        }
+        return crews.get(name);
+    }
```
`crews.get(name)`의 호출을 한번만 해보도록 수정해볼까요?

asebn1 commented on src/main/java/domain/AttendanceStatistics.java:
```diff
@@ -0,0 +1,57 @@
+            TimeAndStatus status = records.findByDate(startDate);
+            startDate = startDate.plusDays(1);
+
+            if (!Holiday.isHoliday(startDate.minusDays(1))) {
+                if (status == null || status.getStatus() == null) {
+                    absenceCount++;
+                    continue;
+                }
+
+                if (status.getStatus().equals("출석")) {
+                    attendanceCount++;
+                }
+                if (status.getStatus().equals("지각")) {
+                    latenessCount++;
+                }
+                if (status.getStatus().equals("결석")) {
+                    absenceCount++;
+                }
+            }
```
[출석, 지각, 결석] 상태가 존재하네요! 
어떻게 개선할 수 있을까요?

eunseongu replied:
AttendanceStatus라는 enum을 생성해서 리팩터링했습니다!

asebn1 commented on src/main/java/domain/AttendanceStatistics.java:
```diff
@@ -0,0 +1,57 @@
+            StatisticsResult statisticsResult = AttendanceStatistics.countStatus(nowDate,
+                crews.get(crewName));
+
+            Penalty penalty = statisticsResult.getPenalty();
+            if (penalty != Penalty.NONE) {
+                result.put(crewName, statisticsResult);
+            }
```
```suggestion
            if (statisticsResult.hasPenalty()) {
                result.put(crewName, statisticsResult);
            }
```

- 값을 가져와서 검증
- 객체에게 검증로직 호출

어떤 차이가 있을까요?

eunseongu replied:
시간이 부족해서 getter를 사용하게 됐는데, 사용하면서도 개선 방법이 존재할 것이라는 생각이 들었습니다.

객체에게 검증로직을 호출하도록 하면 객체가 스스로 검증해서 결과만 알려주게 되어 getter를 사용하지 않도록 개선 가능해보입니다! 리팩터링에 반영했습니다 😊

asebn1 replied:
참고하면 좋을 자료 공유드립니다~

- https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/

asebn1 commented on src/main/java/domain/Holiday.java:
```diff
@@ -0,0 +1,35 @@
+        this.day = day;
+    }
+
+    public static boolean isHoliday(LocalDate localDate) {
```
이 함수도 같은 이유로 개선할 수 있을 것 같아요~

eunseongu replied:
> - 값을 가져와서 검증
> - 객체에게 검증로직 호출

이 이유를 말씀하시는 걸까요?

asebn1 replied:
네네!

asebn1 commented on src/main/java/domain/Holiday.java:
```diff
@@ -0,0 +1,35 @@
+                return true;
+            }
+        }
+        return false;
+    }
+
+}
```
```suggestion
        return false;
    }
}
```

의미 없는 개행은 지양해보아요

asebn1 commented on src/main/java/domain/Records.java:
```diff
@@ -0,0 +1,65 @@
+        return timeAndStatus;
+    }
+
+    public boolean isSameDate(LocalDateTime localDateTime) {
+        LocalDate localDate = localDateTime.toLocalDate();
+        return records.containsKey(localDate);
+    }
```
has는 어떨까요?

eunseongu replied:
넵 언급감사합니다! 리팩터링에 반영했습니다.

asebn1 commented on src/test/java/AttendanceRecordCheckTest.java:
```diff
@@ -0,0 +1,115 @@
+        attendanceManager.createCrew(name, records);
+    }
+
+    @DisplayName("크루 출석 기록을 전날까지 조회한다.")
+    @Test
+    void 출석_기록_조회() {
+        String name = "빙티";
+        Records expectedRecords = attendanceManager.findByName(name);
+        assertThat(expectedRecords.getAttendanceCount()).isEqualTo(records.size());
+    }
+
+    @DisplayName("출석 통계를 정확하게 계산한다.")
+    @Test
+    void 출석_통계_계산() {
+        String name = "빙티";
+        Records records = attendanceManager.findByName(name);
```
`DisplayName`에 서술을 해주어서 함수명은 영어로 작성하는게 적절할 것 같아요

eunseongu replied:
should-when 방식을 사용하여 영어로 변경해 보았습니다!

asebn1 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,145 @@
+import view.InputView;
+import view.OutputView;
+
+public class AttendanceController {
+
```
`controller`는 어떤 목적의 클래스일까요?

eunseongu replied:
이전 코드에서는 여러 역할을 하는 모호한 느낌이 있어서 리팩토링에서 최대한 수정해보았습니다.

- `InputView`를 사용해 사용자 입력을 받고, 입력에 맞는 ` Function`을 동작 시키고 사용자 입력값(크루 닉네임 존재여부)에 대한 검증
- `attendanceManager`가 각 기능을 처리할 수 있도록 하는 역할
- OutputView에서 필요한 데이터를 처리하고 출력하는 역할

세 가지가 존재하도록 변경해보았습니다.

어떤 부분을 더 개선하면 좋을까요?

asebn1 replied:
중요한 부분이라 리마인드차 드린 코멘트였어요 :)

asebn1 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,145 @@
+
+    private final int DEFAULT_YEAR = 2024;
+    private final int DEFAULT_MONTH = 12;
+    private final int DEFAULT_DAY = 13;
```
상수명만 보아서는 2024, 12, 13인지 잘 모를 것 같아요

eunseongu replied:
임의로 날짜를 고정해두고 쓰기 위해 작성한 상수라서 원하는 일자에 맞도록 변경해서 쓸 수 있다는 의미를 담으려고 했습니다!

해당 날짜를 기준으로 출석 과정이 이루어지기 때문에 `ATTENDANCE_` 로 변경해보았는데 명확한 의미가 담겼는지 확신이 들지 않네요. 🥲

asebn1 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,145 @@
+            Records records = attendanceManager.findByName(name);
+
+            StatisticsResult statisticsResult = AttendanceStatistics.countStatus(localDate,
+                records);
```
줄바꿈을 안해도 될 것 같아요

asebn1 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,145 @@
+        return LocalDateTime.parse(dateForm + " " + timeForm, formatter);
+    }
+
+}
```
```suggestion
    }
}
```

여기도 의미없는 개행이 존재하네요!

asebn1 reviewed (changes requested):
하루 안녕하세요 :)
이번 출선구현미션 리뷰어 제로입니다.
이번 미션도 화이팅이에요 🔥🔥🔥
코멘트 확인 부탁드립니다

---

> 한 크루의 상태만 출력하고 싶은 경우에도 항상 여러 객체를 거쳐 값을 반환받고, 최종적으로 Controller가 View에게 넘겨주고 있는데, 이 방식이 괜찮은 것인가에 대한 의문이 들었습니다.

현재 AttendanceManager가 모든 크루의 출석 정보를 직접 관리하는 구조라, 개별 크루의 상태를 조회할 때도 반드시 매니저를 거쳐야 하는 문제가 있을 것 같아요.
도메인 구조를 어떻게 가져갈지 한번 고민해봅시다 :) 
Crew 도메인을 고려해보는 것도 좋은 방향일 것 같아요. 

추가로 일급 컬렉션은 무엇일까요? 요구사항에 `일급 컬렉션을 쓴다.`가 있습니다. 반영해볼까요?

eunseongu committed:
refactor: 상수에 필요한 static과 final을 적용하고 불필요한 개행 삭제

eunseongu committed:
refactor: DateTimeFormatter가 반복 생성되지 않도록 변경

eunseongu committed:
refactor: 반복적으로 호출되는 항목 제거

eunseongu committed:
style: 불필요한 줄바꿈 제거

eunseongu committed:
refactor: 컨트롤러 객체 수정

- 구체적인 표현을 위해 변수명을 변경한다.
- 반복되는 Formatter 생성을 막기 위해 상수로 선언한다.
- 반복 사용 되는 콜론을 상수로 선언한다.
- 가독성 향상을 위해 기능 수행 메서드를 switch-case문으로 변경한다.
- 코드 개선을 위해 중복되는 포맷팅 처리를 제거하고, 메서드로 분리한다.

eunseongu committed:
refactor: 기능 실행 로직을 Function 객체로 추출

- indent 1 이하, 메서드 10줄이하 제한 조건을 만족하기 위해 수정한다.

eunseongu committed:
refactor: 출석 상태를 enum 객체로 관리

eunseongu committed:
refactor: 객체에게 검증로직 호출하도록 수정

eunseongu committed:
refactor: 일급컬렉션을 사용하도록 수정

eunseongu committed:
feat: 휴일인 경우 예외 처리

eunseongu committed:
refactor: 휴일인 경우에 대한 테스트 케이스 추가

eunseongu committed:
feat: 입력 문자열이 비어있는 경우 예외 처리

eunseongu committed:
refactor: 크루 정보가 존재하는지 확인하는 방식 변경

eunseongu committed:
refactor: 테스트 메서드명 영어로 수정

eunseongu committed:
feat: 숫자가 아닌 경우, 날짜 범위가 아닌 경우에 대한 예외 처리

eunseongu committed:
feat: 시간 형식이 유효하지 않은 경우 예외 처리

eunseongu committed:
feat: 캠퍼스 운영시간이 아닌 경우 예외 처리

eunseongu committed:
refactor: 가독성 향상을 위한 컨벤션 통일 및 메서드 분리

eunseongu committed:
refactor: 가독성 향상을 위한 메서드 분리 및 불필요한 메서드 호출 제거

asebn1 commented on src/main/java/controller/FileController.java:
```diff
@@ -0,0 +1,60 @@
+        Map<String, List<LocalDateTime>> result = new HashMap<>();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+
+        while (attendaceScanner.hasNextLine()) {
+            String[] attr = attendaceScanner.nextLine().split(",");
+            String name = attr[0];
+            LocalDateTime dateTime = LocalDateTime.parse(attr[1], formatter);
```
0, 1의 의미를 알 수 없는 것 같아요.

asebn1 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,28 @@
+
+    public String getDescription(){
+        return status;
+    }
+}
+
+
```
개행이 더 들어가있네요 ㅎㅎ

asebn1 commented on src/main/java/domain/Function.java:
```diff
@@ -0,0 +1,24 @@
+        this.functionNumber = functionNumber;
+    }
+
+    public static Function from(String functionNumber) {
+        return Arrays.stream(values())
+                .filter(f->f.functionNumber.equals(functionNumber))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("유효하지 않은 번호입니다."));
+    }
```
인텔리제이 자동정렬 기능을 사용해볼까요? 
```suggestion
    public static Function from(String functionNumber) {
        return Arrays.stream(values())
                .filter(f -> f.functionNumber.equals(functionNumber))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("유효하지 않은 번호입니다."));
    }
```

이런식으로 자동으로 정렬될 것 같아요.

asebn1 commented on src/main/java/controller/FileController.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Map;
+import java.util.Scanner;
+
+public class FileController {
+
+    private static final String FILE_ERROR_MESSAGE = "파일을 불러오는 데 문제가 발생했습니다. 잠시 후 다시 시도해 주세요.";
+
+    private final AttendanceManager attendanceManager;
+
+    public FileController(AttendanceManager attendanceManager) {
+        this.attendanceManager = attendanceManager;
+    }
```
FileController가 MVC 모델 관점에서 보았을 때 `Controller`의 역할과 거리가 먼 것 같아요 🥲

asebn1 reviewed (approved):
하루 안녕하세요
1단 미션 고생 많으셨습니다 :)
간단한 코멘트 남겼는데 2단계 제출 시에 함께 반영해주세요
이번 단계는 이만 머지할게요 
2단계 미션도 화이팅!
